transports/quic/src/hole_punching.rs

```diff
@@ -0,0 +1,126 @@
+use std::{
+    collections::HashMap,
+    net::SocketAddr,
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+    time::Duration,
+};
+
+use futures::{
+    channel::oneshot,
+    future::{Fuse, FusedFuture},
+    prelude::*,
+};
+use futures_timer::Delay;
+use libp2p_identity::PeerId;
+use rand::{distributions, Rng};
+
+use crate::{
+    endpoint::{self, ToEndpoint},
+    Connecting, Connection, Error,
+};
+
+pub(crate) type HolePunchMap =
+    Arc<Mutex<HashMap<(SocketAddr, PeerId), oneshot::Sender<(PeerId, Connection)>>>>;
+
+pub(crate) struct MaybeHolePunchedConnection {
```

Comment:
```suggestion
/// An upgrading inbound QUIC connection that is either a normal inbound connection or an inbound connection corresponding to an in-progress outbound hole punching connection.
pub(crate) struct MaybeHolePunchedConnection {
```

Would make it easier to understand. Don't feel strongly about the concrete wording.